targets/psa_switch/externs/psa_internetChecksum.cpp

```diff
@@ -0,0 +1,119 @@
+/* Copyright 2020-present Cornell University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Yunhe Liu (yunheliu@cs.cornell.edu)
+ *
+ */
+
+#include "psa_internetChecksum.h"
+#include <bm/bm_sim/logger.h>
+#include <execinfo.h>
+
+#define SUM_LENGTH 16
+
+namespace bm {
+
+namespace psa {
+
+uint16_t
+ones_complement_sum(uint16_t x, uint16_t y) {
+  uint32_t _x, _y, s;
+  uint16_t ret;
+  _x = x;
+  _y = y;
+  s = _x + _y;
+  if (s >> 16 == 1) {
+    s = s + 1;
+  }
+  ret = s;
+  return ret;
+}
+
+void
+PSA_InternetChecksum::init() {
+  sum = 0;
+}
+
+void
+PSA_InternetChecksum::clear() {
+  sum = 0;
+}
+
+void
+PSA_InternetChecksum::add(const Field &field) {
+  int len = field.get_nbits();
+  // TODO give a error if len is not a multiple of 16
+  Data field_val(field);
+
+  int i = 0;
+  while (i < len) {
+    Data tmp(field_val);
+    Data mask(0xffff);
+    tmp.bit_and(tmp, mask);
+    uint16_t d = tmp.get<uint16_t>();
+    sum = ones_complement_sum(sum, d);
+    field_val.bm::Data::shift_right(field_val, SUM_LENGTH);
+    i += SUM_LENGTH;
```

Comment:
that looks like a very inefficient implementation, I don't understand why this has to be done bit by bit

Reply:
When you say bit by bit, you mean that it is doing it in 16-bit word chunks?

I have no argument that there might be more efficient ways to implement it.  In practice I suspect this function will be called with lists of fields that are typically no more than 18 bytes (for IPv4 headers, minus checksum) or 18+2*16+2 = 42 bytes for TCP headers with IPv6 pseudo-headers, so having a correct implementation with perhaps a TODO for considering performance measurement and enhancement might be appropriate?

Reply:
actually it's my bad, I read this too fast. I saw the while loop and didn't see that the increment was `i += SUM_LENGTH` at the end. Maybe we can replace it with `for (int i = 0; i < len; i+= SUM_LENGTH)`. It's less verbose and less confusing to me :)

on an unrelated node, maybe @yunheL can explain why the explicit reference to `bm::Data::shift_right` is needed in `field_val.bm::Data::shift_right(field_val, SUM_LENGTH);`? `field_val` is already a `Data` object, and even if it was a `Field`, I believe it sill would not be required.

Reply:
However, regarding implementation performance, we should avoid creating all these temporary `Data` objects, which are kind of expensive. The internal state (`sum`) is already a uint16, so we can extract 16-bit chunks from the input `field` and do all the arithmetic on that. Looking at the current code, I only see 16-bit arithmetic, so there is no need for arbitrary-precision arithmetic / big nums.

That's one more reason for adding some C++ unit tests. We can quickly modify them to act as a benchmark.

Reply:
And to clarify "we can extract 16-bit chunks from the input `field`", I think potentially the best is to get the binary string for the input `field` and read it 2 bytes at a time.